File: LeetCode/101-symmetric-tree/Solution.java
/**
 * Solution class to check if a binary tree is symmetric.
 */
class Solution {

  /**
   * Determines whether a binary tree is symmetric around its center.
   *
   * @param root The root of the binary tree.
   * @return True if the tree is symmetric, false otherwise.
   */
  public boolean isSymmetric(TreeNode root) {
    // An empty tree is symmetric
    if (root == null) {
      return true;
    }
    // Check if the left and right subtrees are mirror reflections
    return isMirror(root.left, root.right);
  }

  /**
   * Auxiliary function to check if two subtrees are mirror reflections of each other.
   *
   * @param left The root of the left subtree.
   * @param right The root of the right subtree.
   * @return True if the two subtrees are mirrors, false otherwise.
   */
  private boolean isMirror(TreeNode left, TreeNode right) {
    // Base cases
    // Both subtrees are empty
    if (left == null && right == null) {
      return true;
    }

    // One subtree is empty
    if (left == null || right == null) {
      return false;
    }

    // Root values of the subtrees do not match
    if (left.val != right.val) {
      return false;
    }

    // Recursively check if:
    // 1. The left child of the left subtree is a mirror of the right child of the right subtree
    // 2. The right child of the left subtree is a mirror of the left child of the right subtree
    return isMirror(left.left, right.right) && isMirror(left.right, right.left);
  }
}

File: LeetCode/101-symmetric-tree/solution.ts
/**
 * Definition for a binary tree node.
 */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * Determines whether a binary tree is symmetric around its center.
 *
 * @param root - The root of the binary tree.
 * @returns True if the tree is symmetric, false otherwise.
 */
function isSymmetric(root: TreeNode | null): boolean {
  if (root === null) {
      return true; // An empty tree is symmetric
  }
  return isMirror(root.left, root.right);
}

/**
* Helper function that checks whether two subtrees are mirror reflections of each other.
*
* @param left - The root of the left subtree.
* @param right - The root of the right subtree.
* @returns True if the subtrees are mirror reflections, false otherwise.
*/
function isMirror(left: TreeNode | null, right: TreeNode | null): boolean {
  // Base cases
  if (left === null && right === null) {
      return true; // Both subtrees are empty
  }
  if (left === null || right === null) {
      return false; // Only one subtree is empty
  }
  if (left.val !== right.val) {
      return false; // Values of the nodes do not match
  }

  // Recursively check for mirror symmetry
  return isMirror(left.left, right.right) && isMirror(left.right, right.left);
}